<?xml version="1.0" encoding="utf-8" ?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="Docutils 0.12: http://docutils.sourceforge.net/" />
<title></title>
<style type="text/css">

/*
:Author: David Goodger (goodger@python.org)
:Id: $Id: html4css1.css 7614 2013-02-21 15:55:51Z milde $
:Copyright: This stylesheet has been placed in the public domain.

Default cascading style sheet for the HTML output of Docutils.

See http://docutils.sf.net/docs/howto/html-stylesheets.html for how to
customize this style sheet.
*/

/* used to remove borders from tables and images */
.borderless, table.borderless td, table.borderless th {
  border: 0 }

table.borderless td, table.borderless th {
  /* Override padding for "table.docutils td" with "! important".
     The right padding separates the table cells. */
  padding: 0 0.5em 0 0 ! important }

.first {
  /* Override more specific margin styles with "! important". */
  margin-top: 0 ! important }

.last, .with-subtitle {
  margin-bottom: 0 ! important }

.hidden {
  display: none }

a.toc-backref {
  text-decoration: none ;
  color: black }

blockquote.epigraph {
  margin: 2em 5em ; }

dl.docutils dd {
  margin-bottom: 0.5em }

object[type="image/svg+xml"], object[type="application/x-shockwave-flash"] {
  overflow: hidden;
}

/* Uncomment (and remove this text!) to get bold-faced definition list terms
dl.docutils dt {
  font-weight: bold }
*/

div.abstract {
  margin: 2em 5em }

div.abstract p.topic-title {
  font-weight: bold ;
  text-align: center }

div.admonition, div.attention, div.caution, div.danger, div.error,
div.hint, div.important, div.note, div.tip, div.warning {
  margin: 2em ;
  border: medium outset ;
  padding: 1em }

div.admonition p.admonition-title, div.hint p.admonition-title,
div.important p.admonition-title, div.note p.admonition-title,
div.tip p.admonition-title {
  font-weight: bold ;
  font-family: sans-serif }

div.attention p.admonition-title, div.caution p.admonition-title,
div.danger p.admonition-title, div.error p.admonition-title,
div.warning p.admonition-title, .code .error {
  color: red ;
  font-weight: bold ;
  font-family: sans-serif }

/* Uncomment (and remove this text!) to get reduced vertical space in
   compound paragraphs.
div.compound .compound-first, div.compound .compound-middle {
  margin-bottom: 0.5em }

div.compound .compound-last, div.compound .compound-middle {
  margin-top: 0.5em }
*/

div.dedication {
  margin: 2em 5em ;
  text-align: center ;
  font-style: italic }

div.dedication p.topic-title {
  font-weight: bold ;
  font-style: normal }

div.figure {
  margin-left: 2em ;
  margin-right: 2em }

div.footer, div.header {
  clear: both;
  font-size: smaller }

div.line-block {
  display: block ;
  margin-top: 1em ;
  margin-bottom: 1em }

div.line-block div.line-block {
  margin-top: 0 ;
  margin-bottom: 0 ;
  margin-left: 1.5em }

div.sidebar {
  margin: 0 0 0.5em 1em ;
  border: medium outset ;
  padding: 1em ;
  background-color: #ffffee ;
  width: 40% ;
  float: right ;
  clear: right }

div.sidebar p.rubric {
  font-family: sans-serif ;
  font-size: medium }

div.system-messages {
  margin: 5em }

div.system-messages h1 {
  color: red }

div.system-message {
  border: medium outset ;
  padding: 1em }

div.system-message p.system-message-title {
  color: red ;
  font-weight: bold }

div.topic {
  margin: 2em }

h1.section-subtitle, h2.section-subtitle, h3.section-subtitle,
h4.section-subtitle, h5.section-subtitle, h6.section-subtitle {
  margin-top: 0.4em }

h1.title {
  text-align: center }

h2.subtitle {
  text-align: center }

hr.docutils {
  width: 75% }

img.align-left, .figure.align-left, object.align-left {
  clear: left ;
  float: left ;
  margin-right: 1em }

img.align-right, .figure.align-right, object.align-right {
  clear: right ;
  float: right ;
  margin-left: 1em }

img.align-center, .figure.align-center, object.align-center {
  display: block;
  margin-left: auto;
  margin-right: auto;
}

.align-left {
  text-align: left }

.align-center {
  clear: both ;
  text-align: center }

.align-right {
  text-align: right }

/* reset inner alignment in figures */
div.align-right {
  text-align: inherit }

/* div.align-center * { */
/*   text-align: left } */

ol.simple, ul.simple {
  margin-bottom: 1em }

ol.arabic {
  list-style: decimal }

ol.loweralpha {
  list-style: lower-alpha }

ol.upperalpha {
  list-style: upper-alpha }

ol.lowerroman {
  list-style: lower-roman }

ol.upperroman {
  list-style: upper-roman }

p.attribution {
  text-align: right ;
  margin-left: 50% }

p.caption {
  font-style: italic }

p.credits {
  font-style: italic ;
  font-size: smaller }

p.label {
  white-space: nowrap }

p.rubric {
  font-weight: bold ;
  font-size: larger ;
  color: maroon ;
  text-align: center }

p.sidebar-title {
  font-family: sans-serif ;
  font-weight: bold ;
  font-size: larger }

p.sidebar-subtitle {
  font-family: sans-serif ;
  font-weight: bold }

p.topic-title {
  font-weight: bold }

pre.address {
  margin-bottom: 0 ;
  margin-top: 0 ;
  font: inherit }

pre.literal-block, pre.doctest-block, pre.math, pre.code {
  margin-left: 2em ;
  margin-right: 2em }

pre.code .ln { color: grey; } /* line numbers */
pre.code, code { background-color: #eeeeee }
pre.code .comment, code .comment { color: #5C6576 }
pre.code .keyword, code .keyword { color: #3B0D06; font-weight: bold }
pre.code .literal.string, code .literal.string { color: #0C5404 }
pre.code .name.builtin, code .name.builtin { color: #352B84 }
pre.code .deleted, code .deleted { background-color: #DEB0A1}
pre.code .inserted, code .inserted { background-color: #A3D289}

span.classifier {
  font-family: sans-serif ;
  font-style: oblique }

span.classifier-delimiter {
  font-family: sans-serif ;
  font-weight: bold }

span.interpreted {
  font-family: sans-serif }

span.option {
  white-space: nowrap }

span.pre {
  white-space: pre }

span.problematic {
  color: red }

span.section-subtitle {
  /* font-size relative to parent (h1..h6 element) */
  font-size: 80% }

table.citation {
  border-left: solid 1px gray;
  margin-left: 1px }

table.docinfo {
  margin: 2em 4em }

table.docutils {
  margin-top: 0.5em ;
  margin-bottom: 0.5em }

table.footnote {
  border-left: solid 1px black;
  margin-left: 1px }

table.docutils td, table.docutils th,
table.docinfo td, table.docinfo th {
  padding-left: 0.5em ;
  padding-right: 0.5em ;
  vertical-align: top }

table.docutils th.field-name, table.docinfo th.docinfo-name {
  font-weight: bold ;
  text-align: left ;
  white-space: nowrap ;
  padding-left: 0 }

/* "booktabs" style (no vertical lines) */
table.docutils.booktabs {
  border: 0px;
  border-top: 2px solid;
  border-bottom: 2px solid;
  border-collapse: collapse;
}
table.docutils.booktabs * {
  border: 0px;
}
table.docutils.booktabs th {
  border-bottom: thin solid;
  text-align: left;
}

h1 tt.docutils, h2 tt.docutils, h3 tt.docutils,
h4 tt.docutils, h5 tt.docutils, h6 tt.docutils {
  font-size: 100% }

ul.auto-toc {
  list-style-type: none }

</style>
</head>
<body>
<div class="document">


<img alt="bcbio banner" class="align-center" src="artwork/github.png" />
<p>Validated, scalable, community developed variant calling and RNA-seq analysis.
You write a high level configuration file specifying your inputs and analysis
parameters.  This input drives a parallel run that handles distributed
execution, idempotent processing restarts and safe transactional steps.
bcbio-nextgen provides a shared community resource that handles the data processing
component of sequencing analysis, providing researchers with more time to focus
on the downstream biology.</p>
<div class="section" id="features">
<h1>Features</h1>
<ul class="simple">
<li>Community developed: We welcome contributors with the goal of
overcoming the biological, algorithmic and computational challenges
that face individual developers working on complex pipelines in
quickly changing research areas. See our <a class="reference external" href="https://bcbio-nextgen.readthedocs.org/en/latest/contents/introduction.html#users">users page</a> for examples
of bcbio-nextgen deployments, and the <a class="reference external" href="https://bcbio-nextgen.readthedocs.org/en/latest/contents/code.html">developer documentation</a> for
tips on contributing.</li>
<li>Installation: <a class="reference external" href="https://bcbio-nextgen.readthedocs.org/en/latest/contents/installation.html#automated">A single installer script</a> prepares all
third party software, data libraries and system configuration files.</li>
<li><a class="reference external" href="http://bcbio.wordpress.com/2014/05/12/wgs-trio-variant-evaluation/">Automated validation</a>: Compare variant calls against common reference
materials or sample specific SNP arrays to ensure call correctness.
Incorporation of multiple approaches for alignment, preparation and
variant calling enable unbiased comparisons of algorithms.</li>
<li>Distributed: Focus on <a class="reference external" href="http://bcbio.wordpress.com/2013/05/22/scaling-variant-detection-pipelines-for-whole-genome-sequencing-analysis/">parallel analysis and scaling</a> to handle
large population studies and whole genome analysis. Runs on single
multicore computers, in compute clusters using <a class="reference external" href="http://ipython.org/ipython-doc/dev/index.html">IPython parallel</a>,
or on the Amazon cloud. See the <a class="reference external" href="https://bcbio-nextgen.readthedocs.org/en/latest/contents/parallel.html">parallel documentation</a> for full
details.</li>
<li>Multiple analysis algorithms: bcbio-nextgen provides configurable
<a class="reference external" href="https://bcbio-nextgen.readthedocs.org/en/latest/contents/pipelines.html">variant calling and RNA-seq pipelines</a>.</li>
</ul>
</div>
<div class="section" id="quick-start">
<h1>Quick start</h1>
<ol class="arabic">
<li><p class="first"><a class="reference external" href="https://bcbio-nextgen.readthedocs.org/en/latest/contents/installation.html#automated">Install</a> <tt class="docutils literal"><span class="pre">bcbio-nextgen</span></tt> with all tool dependencies and data files:</p>
<pre class="literal-block">
wget https://raw.github.com/chapmanb/bcbio-nextgen/master/scripts/bcbio_nextgen_install.py
python bcbio_nextgen_install.py /usr/local/share/bcbio --tooldir=/usr/local \
  --genomes GRCh37 --aligners bwa --aligners bowtie2
</pre>
<p>producing an editable <a class="reference external" href="https://github.com/chapmanb/bcbio-nextgen/blob/master/config/bcbio_system.yaml">system configuration file</a> referencing the installed
software, data and system information.</p>
</li>
<li><p class="first"><a class="reference external" href="https://bcbio-nextgen.readthedocs.org/en/latest/contents/configuration.html#automated-sample-configuration">Automatically create a processing description</a> of sample FASTQ and BAM files
from your project, and a CSV file of sample metadata:</p>
<pre class="literal-block">
bcbio_nextgen.py -w template freebayes-variant project1.csv sample1.bam sample2_1.fq sample2_2.fq
</pre>
<p>This produces a <a class="reference external" href="https://github.com/chapmanb/bcbio-nextgen/blob/master/config/bcbio_sample.yaml">sample description file</a> containing pipeline <a class="reference external" href="https://bcbio-nextgen.readthedocs.org/en/latest/contents/configuration.html">configuration options</a>.</p>
</li>
<li><p class="first">Run analysis, distributed across 8 local cores:</p>
<pre class="literal-block">
cd project1/work
bcbio_nextgen.py ../config/project1.yaml -n 8
</pre>
</li>
</ol>
</div>
<div class="section" id="documentation">
<h1>Documentation</h1>
<p>See the <a class="reference external" href="https://bcbio-nextgen.readthedocs.org">full documentation at ReadTheDocs</a>. We welcome enhancements
or problem reports using <a class="reference external" href="https://github.com/chapmanb/bcbio-nextgen/issues">GitHub</a> and discussion on the
<a class="reference external" href="https://groups.google.com/d/forum/biovalidation">biovalidation mailing list</a>.</p>
</div>
<div class="section" id="contributors">
<h1>Contributors</h1>
<ul class="simple">
<li><a class="reference external" href="https://github.com/mjafin">Miika Ahdesmaki</a>, AstraZeneca</li>
<li><a class="reference external" href="https://github.com/lbeltrame">Luca Beltrame</a>, IRCCS &quot;Mario Negri&quot; Institute for Pharmacological Research, Milan, Italy</li>
<li><a class="reference external" href="https://github.com/abushoy">Alla Bushoy</a>, AstraZeneca</li>
<li><a class="reference external" href="https://github.com/guillermo-carrasco">Guillermo Carrasco</a>, Science for Life Laboratory, Stockholm</li>
<li><a class="reference external" href="http://www.simonsfoundation.org/about-us/staff/staff-bios/#nick-carriero-ph-d">Nick Carriero</a>, Simons Foundation</li>
<li><a class="reference external" href="https://github.com/chapmanb">Brad Chapman</a>, Harvard Chan Bioinformatics Core</li>
<li><a class="reference external" href="https://github.com/saketkc">Saket Choudhary</a>, University Of Southern California</li>
<li><a class="reference external" href="https://github.com/peterjc">Peter Cock</a>, The James Hutton Institute</li>
<li><a class="reference external" href="https://github.com/matted">Matt Edwards</a>, MIT</li>
<li><a class="reference external" href="https://github.com/mariogiov">Mario Giovacchini</a>, Science for Life Laboratory, Stockholm</li>
<li><a class="reference external" href="https://twitter.com/kgutwin">Karl Gutwin</a>, Biogen</li>
<li><a class="reference external" href="https://github.com/hammer">Jeff Hammerbacher</a>, Icahn School of Medicine at Mount Sinai</li>
<li><a class="reference external" href="https://github.com/kern3020">John Kern</a></li>
<li><a class="reference external" href="https://github.com/roryk">Rory Kirchner</a>, Harvard Chan Bioinformatics Core</li>
<li><a class="reference external" href="https://github.com/jsnowacki">Jakub Nowacki</a>, AstraZeneca</li>
<li><a class="reference external" href="https://github.com/jwm">John Morrissey</a>, Harvard Chan Bioinformatics Core</li>
<li><a class="reference external" href="https://github.com/lpantano">Lorena Pantano</a>, Harvard Chan Bioinformatics Core</li>
<li><a class="reference external" href="https://github.com/brentp">Brent Pedersen</a>, University of Colorado Denver</li>
<li><a class="reference external" href="https://github.com/porterjamesj">James Porter</a>, The University of Chicago</li>
<li><a class="reference external" href="https://github.com/vals">Valentine Svensson</a>, Science for Life Laboratory, Stockholm</li>
<li><a class="reference external" href="https://github.com/tanglingfung">Paul Tang</a>, UCSF</li>
<li><a class="reference external" href="https://github.com/brainstorm">Roman Valls</a>, Science for Life Laboratory, Stockholm</li>
<li><a class="reference external" href="https://github.com/kevyin">Kevin Ying</a>, Garvan Institute of Medical Research, Sydney, Australia</li>
</ul>
</div>
<div class="section" id="license">
<h1>License</h1>
<p>The code is freely available under the <a class="reference external" href="http://www.opensource.org/licenses/mit-license.html">MIT license</a>.</p>
</div>
</div>
</body>
</html>
