![bcbio banner](contents/images/banner.png)

A python toolkit providing best-practice pipelines for fully automated high throughput sequencing analysis. You write a high level configuration file specifying your inputs and analysis parameters. This input drives a parallel pipeline that handles distributed execution, idempotent processing restarts and safe transactional steps. The goal is to provide a shared community resource that handles the data processing component of sequencing analysis, providing researchers with more time to focus on the downstream biology.

## Contents

* [Getting started](contents/intro)
* [Single cell RNA-seq](contents/single_cell)
* [Bulk RNA-seq expression](contents/bulk_rnaseq)
* [Bulk RNA-seq variants](contents/rnaseq_variants)
* [3prime digital gene expression](contents/3prime_dge)
* [ATAC-seq](contents/atac)
* [Introduction](contents/about)
* [Installation](contents/installation)
* [Pipelines](contents/pipelines)
* [Configuration](contents/configuration)
* [Parallel execution](contents/parallel)
* [Common Workflow Language (CWL)](contents/cwl)
* [Cloud](contents/cloud)
* [Outputs](contents/outputs)
* [Development](contents/development)
* [Internals](contents/internals)
* [Presentations](contents/presentations)
* [Teaching](contents/teaching)
* [Citations](contents/citations)
